lib/yield_star_client/errors.rb

```diff
@@ -17,39 +17,50 @@ def initialize(message=nil, code=nil)
     #
     # @param [Savon::SOAP::Fault] soap_error the error object raised by the soap client
     # @return [YieldStarClient::ServerError] the corresponding YieldStarClient error
-    def self.translate_fault(soap_error)
-      if soap_error.instance_of? Savon::HTTPError
-        fault = soap_error.to_hash
+    def self.translate_fault(error)
+      if error.instance_of? Savon::HTTPError
+        http_build_error_object(error)
+      elsif error.instance_of? Savon::SOAPFault
+        soap_build_error_object(error)
+      end
+    end
 
-        if fault[:code] == 401
-          error_class = YieldStarClient::AuthenticationError
-          message = "Authentication Error"
-        else
-          error_class = YieldStarClient::ServerError
-          message = fault[:body]
-        end
+    private
 
-        fault_detail = {:code => fault[:code], :message => message}
-      elsif soap_error.instance_of? Savon::SOAPFault
-        fault = soap_error.to_hash[:fault]
-        error_class = YieldStarClient::ServerError
-        fault_detail = {:code => fault[:faultcode], :message => fault[:faultstring]}
+    def self.http_build_error_object(http_error)
+      http_error_hash = http_error.to_hash
+      code = http_error_hash[:code]
+      body = http_error_hash[:body]
 
-        if detail = fault[:detail]
-          if detail.has_key?(:authentication_fault)
-            error_class = YieldStarClient::AuthenticationError
-            fault_detail = detail[:authentication_fault]
-          elsif detail.has_key?(:operation_fault)
-            error_class = YieldStarClient::OperationError
-            fault_detail = detail[:operation_fault]
-          elsif detail.has_key?(:internal_error_fault)
-            error_class = YieldStarClient::InternalError
-            fault_detail = detail[:internal_error_fault]
-          end
-        end
+      case code
+      when 401
+        YieldStarClient::AuthenticationError.new("Authentication Error", code)
+      else
+        YieldStarClient::ServerError.new(body, code)
       end
+    end
 
-      error_class.new(fault_detail[:message], fault_detail[:code])
+    def self.soap_build_error_object(soap_error)
+      soap_error_hash = soap_error.to_hash[:fault]
+      detail = soap_error_hash[:detail]
+
```

Comment:
You can move the `else` condition here and do an early return. This way, you won't need to have two conditions for each `if` or `elsif`